File: src/main/java/de/tuberlin/cit/allocationassistant/Freamon.java
package de.tuberlin.cit.allocationassistant;

import akka.actor.*;
import com.typesafe.config.ConfigFactory;

import java.io.Serializable;

/**
 * Sends messages to the Freamon MonitorMasterActor
 */
public class Freamon extends UntypedActor {

    private static final ActorRef instance = getActor();

    private static ActorRef getActor() {
        // TODO pass application.conf path in config.xml
        final com.typesafe.config.Config config = ConfigFactory
            .parseString("akka.remote.netty.tcp.hostname=" + Config.getInstance().getConfigItem("akkaHost")
                    + "\nakka.remote.netty.tcp.port=" + Config.getInstance().getConfigItem("akkaPort"))
            .withFallback(ConfigFactory.load());

        final ActorSystem actorSystem = ActorSystem.create("allocationAssistantSystem", config);

        return actorSystem.actorOf(Props.create(Freamon.class), "freamonSender");
    }

    private final ActorSelection freamonMaster = getFreamonMasterActor();

    private ActorSelection getFreamonMasterActor() {
        final Address masterSystemPath = new Address("akka.tcp",
            Config.getInstance().getConfigItem("freamonMasterSystemName"),
            Config.getInstance().getConfigItem("freamonMasterHost"),
            Integer.parseInt(Config.getInstance().getConfigItem("freamonMasterPort")));

        final String masterActorPath = masterSystemPath.toString() + "/user/"
            + Config.getInstance().getConfigItem("freamonMasterActorName");

        return getContext().actorSelection(masterActorPath);
    }

    @Override
    public void onReceive(Object message) throws Exception {
        freamonMaster.forward(message, getContext());
    }

    public static void onSubmit(String jobID) {
        // we could use this to add not-yet-started jobs to the db
    }

    public static void onStart(String jobID, long startTime) {
        instance.tell(new Serializable[]{"jobStarted", jobID, startTime}, instance);
    }

    public static void onStop(String jobID, long stopTime) {
        instance.tell(new Serializable[]{"jobStopped", jobID, stopTime}, instance);
    }
}

File: src/main/scala/de/tuberlin/cit/allocationassistant/Freamon.scala
package de.tuberlin.cit.allocationassistant

import akka.actor.{ActorSystem, Address}
import akka.pattern.ask
import akka.util.Timeout
import com.typesafe.config.Config
import de.tuberlin.cit.freamon.api._

import scala.concurrent.Await
import scala.concurrent.duration._

/** Utilities for communicating with Freamon */
class Freamon(config: Config) {
  val freamonMaster = {
    val actorSystemName = config.getString("allocation-assistant.actors.system")
    val actorSystem = ActorSystem(actorSystemName, config)
    val freamonSystemPath = new Address("akka.tcp",
      config.getString("freamon.actors.systems.master.name"),
      config.getString("freamon.hosts.master.hostname"),
      config.getInt("freamon.hosts.master.port"))
    val freamonActorPath = freamonSystemPath.toString + "/user/" + config.getString("freamon.actors.systems.master.actor")

    actorSystem.actorSelection(freamonActorPath)
  }

  def getPreviousRuns(jarWithArgs: String): PreviousRuns = {
    implicit val timeout = Timeout(5 seconds)
    val future = freamonMaster ? FindPreviousRuns(jarWithArgs)
    try {
      Await.result(future, timeout.duration).asInstanceOf[PreviousRuns]
    } catch {
      case e: java.util.concurrent.TimeoutException =>
        println(s"Freamon did not respond with PreviousRuns after $timeout")
        throw e
    }
  }
}
